service/kpi.go

```
@@ -0,0 +1,452 @@
+	}
+
+	var kpi interface{}
+
+	switch kpiType {
+	case "feedback":
+		var feedbackKpi models.FeedbackKpi
+		if err := s.Db.Where("id = ?", kpiID).First(&feedbackKpi).Error; err != nil {
+			if errors.Is(err, gorm.ErrRecordNotFound) {
+				c.JSON(http.StatusNotFound, gin.H{"error": "Feedback KPI not found"})
+				return
+			}
+			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch Feedback KPI"})
+			return
+		}
+		kpi = feedbackKpi
+	case "observatory":
+		var observatoryKpi models.ObservatoryKpi
+		if err := s.Db.Where("id = ?", kpiID).First(&observatoryKpi).Error; err != nil {
+			if errors.Is(err, gorm.ErrRecordNotFound) {
+				c.JSON(http.StatusNotFound, gin.H{"error": "Observatory KPI not found"})
+				return
+			}
+			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch Observatory KPI"})
+			return
+		}
+		kpi = observatoryKpi
+	case "questionnaire":
+		var questionnaireKpi models.QuestionaireKpi
+		if err := s.Db.Where("id = ?", kpiID).First(&questionnaireKpi).Error; err != nil {
+			if errors.Is(err, gorm.ErrRecordNotFound) {
+				c.JSON(http.StatusNotFound, gin.H{"error": "Questionnaire KPI not found"})
+				return
+			}
+			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch Questionnaire KPI"})
+			return
+		}
+		kpi = questionnaireKpi
+	case "measured":
+		var measuredKpi models.MeasuredKpi
+		if err := s.Db.Where("id = ?", kpiID).First(&measuredKpi).Error; err != nil {
+			if errors.Is(err, gorm.ErrRecordNotFound) {
+				c.JSON(http.StatusNotFound, gin.H{"error": "Measured KPI not found"})
+				return
+			}
+			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch Measured KPI"})
+			return
+		}
+		kpi = measuredKpi
+	default:
+		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid KpiType query parameter"})
+		return
+	}
+
+	c.JSON(http.StatusOK, gin.H{"data": kpi})
+}
+
+func (s *KPIService) UpdateKPI(c *gin.Context) {
+	kpiType := c.Query("KpiType")
+	kpiID := c.Param("id")
+
```

It should first check if anoter KPI with the same name exists or not. If yes, then it should return an error.